### ISSUE TITLE
no jvm metrics gathering because of wrong order of Bean registrations

### ISSUE THREAD

**Issue by logemann** [micrometer-metrics/micrometer#823]
Having plugged in spring-boot-starter-actuator and micrometer-registry-graphite into my Spring Boot 2 Application, **the only** metrics i get are those from Hikari Pool. No others like jvm or system ones.

After extensive debugging, i found out that the order of bean registrations might be a problem in my app. Disclaimer: I only added those 2 deps and added 3 lines of spring boot config. So no manual wiring with classes on my side.

The problem in detail seems to be that MeterRegistryPostProcessor @Bean in MetricsAutoConfiguration.class is instanciated before the GraphiteMeterRegistry @Bean in  GraphiteMetricsExportAutoConfiguration.class.

The result is that the PostProcessor is not applied on the GraphiteRegistry and thus no configuration is being done on it.

Looking at the class mentioned class:

```
@Configuration
@AutoConfigureBefore({ CompositeMeterRegistryAutoConfiguration.class,
		SimpleMetricsExportAutoConfiguration.class })
@AutoConfigureAfter(MetricsAutoConfiguration.class)
@ConditionalOnBean(Clock.class)
@ConditionalOnClass(GraphiteMeterRegistry.class)
@ConditionalOnProperty(prefix = "management.metrics.export.graphite", name = "enabled", havingValue = "true", matchIfMissing = true)
@EnableConfigurationProperties(GraphiteProperties.class)
public class GraphiteMetricsExportAutoConfiguration {
...
}
```
There is a @AutoConfigureAfter annotation. What i dont get is, what does that mean, does it mean that this GraphiteMetricsExportAutoConfiguration @Configuration should be run after MetricsAutoConfiguration or vice versa?

Since i dont control any of those classes and i also dont use them programmatically, i dont know what to do on my side. I tried to replace the GraphiteMetricsExportAutoConfiguration with my own implementation but that didnt work. Perhaps i need to fork the micrometer-registry-graphit and play around with some modifiations on that, but have done yet.

**Comment by izeye** [comment]
This looks a duplicate of [this SO question](https://stackoverflow.com/questions/52148543/spring-boot-2-actuator-doesnt-publish-jvm-metric) which I've already answered.

@logemann As I said there, I failed to reproduce your problem with your description on the SO, so if you could make [the sample I provided](https://github.com/izeye/sample-micrometer-spring-boot/tree/graphite) to be broken in the way you're seeing, I can take another look.

**Comment by izeye** [comment]
And as the `MeterRegistryPostProcessor` is a `BeanPostProcessor`, the current ordering doesn't seem to be wrong to me unless I'm missing something.

**Comment by logemann** [comment]
i cant break your example? How should i do that? For some reason nobody knows, my application behaves differently. And to be honest, from a technical point of view, i dont get why the configuration of those registries must be done by a PostProcessor which is also a bean. Why not just put the configuration in the super class of the Registry. This way one there would never be an bean call sequence problem.

But anyway. I am not sure how to proceed. This is all Spring magic happening behind the scenes where i control nothing of it. I can apply the same stuff to another spring boot project of mine and see what happens there.

**Comment by izeye** [comment]
> i cant break your example. How should i do that?

@logemann The sample is minimal and working as expected, so I thought you could trigger the problem by supplying your setup to it.

> But back to my question. Means AutoConfigureAfter that the MetricsAutoConfiguration should run AFTER the current class?

No, the opposite. [Its Javadoc](https://docs.spring.io/spring-boot/docs/current/api/org/springframework/boot/autoconfigure/AutoConfigureAfter.html) says as follows:

> Hint for that an auto-configuration should be applied after other specified auto-configuration classes.

**Comment by logemann** [comment]
my application is at about 90Mb so, no, unfortunately i cant apply my config to the example.

But back to the AutoConfigureAfter. Then the code runs as expected from Spring perspective. But IMO MetricsAutoConfiguration should be run BEFORE the Registries get instanciated shouldnt it? So that all registries that are created and being put into the context will run through the PostProcessor.

I really would like to talk about this fact. Because in my understanding of Spring Contexts, thats the way it should be.

**Comment by izeye** [comment]
> my application is at about 90Mb so, no, unfortunately i cant apply my config to the example.

@logemann Sorry if I wasn't clear but what I wanted with providing a working minimal sample and asking you to make it broken is to get a simple project reproducing the issue. So I didn't mean your whole setup but some minimal setup, especially focused on Micrometer. In my experience, trying to create a minimal sample usually gives a hint on what the cause is regardless it's a bug or a configuration error.

> But back to the AutoConfigureAfter. Then the code runs as expected from Spring perspective. But IMO MetricsAutoConfiguration should be run BEFORE the Registries get instanciated shouldnt it? So that all registries that are created and being put into the context will run through the PostProcessor.
> 
> I really would like to talk about this fact. Because in my understanding of Spring Contexts, thats the way it should be.

I tried to answer your question with the following comment above:

> And as the `MeterRegistryPostProcessor` is a `BeanPostProcessor`, the current ordering doesn't seem to be wrong to me unless I'm missing something.

It doesn't seem to resolve your question in the way what you want. Let me try to rephrase it.

The `MeterRegistryPostProcessor` is free to go anywhere as it's a `BeanPostProcessor`. So `MetricsAutoConfiguration` can go after `GraphiteMetricsExportAutoConfiguration` if `MeterRegistryPostProcessor` is the only member in `MetricsAutoConfiguration` but there is also a `Clock` which `GraphiteMetricsExportAutoConfiguration` depends on, so the current ordering should be preserved.

**Comment by logemann** [comment]
Ok. I ve made progress. I forked your example and checked the callstack of some breakpoints and stuff. First of all i really cant agree on:

> So MetricsAutoConfiguration can go after GraphiteMetricsExportAutoConfiguration

A postprocessor on a context which is just registered cant act on a bean which is already in the context on the registration. Its as simple as that. And all my tests with your example shows me exactly this behavior. Now i checked why (in my app) the `GraphiteMetricsExportAutoConfiguration` gets called before the `MetricsAutoConfiguration`  and somewhere in the callstack i saw a a `HikariDataSource.class`. Now i wondered what the heck this class has to do with `MetricsAutoConfiguration` creation.

 I couldnt figure out exactly because the spring callstack is a nightmare in terms of levels, but i found a `ShiroConfiguration` Class in my app which injects a `DataSource` and use it for a Security Realm. And now it gets weird. This @Configuratiton Class (and its @Bean Methods) was called pretty soon in my bootstrapping process and as soon as the DataSource got injected (and used) it somehow triggered this `GraphiteMetricsExportAutoConfiguration` (via the HikariAutoConfigStuff most likely). I am sure because if i remove this datasource (which would break my app of course), the ordering is perfect again, because the dataSource is not used.

I wanted to put those things into your example (spring-data-jpa dependency and a custom @Configuration class which uses the DataSource) but unfortunately, the custom config class was fired after those configurators in the actuator-spring package. And even with AutoConfigureBefore/After i couldnt change the order of @Configuration Classes.

I have no solution yet, but i am sure that the DataSource is the problem somehow.

**Comment by logemann** [comment]
Ok. Its a miracle that i was able to reproduce it in the test project. Please check out:
https://github.com/logemann/sample-micrometer-spring-boot

You can expirement with the BreakingConfig Class. I added also Shiro as dep to be able to have exactly the same class as in my project. If you remove the @Configuration annotation or the @Bean annotation on the realm method, everything works as expected.

Now its your turn. It took me around 6 hours to get there and this one is _really_ nasty.

**Comment by izeye** [comment]
@logemann Thanks for the sample!

`ShiroFilterFactoryBean` which is a `BeanPostProcessor` and has a dependency on `DefaultWebSecurityManager` will trigger early initialisation. I reworked it to prevent `ShiroFilterFactoryBean` from triggering early initialisation and created https://github.com/logemann/sample-micrometer-spring-boot/pull/1 against your fork.

Closing as this doesn't seem to be a Micrometer issue.

**Comment by andrejleitner** [comment]
Hi folks, I'm probably facing the similar issue (as described in the mentioned [SO question](https://stackoverflow.com/q/52148543)) - seeing only fragment of metrics. I don't use graphite, but yes, there is shiro deployed. I excluded the default config with:
```
@SpringBootApplication(exclude = { ShiroWebFilterConfiguration.class })
```
... and overrode it with:
```
@Configuration
@ConditionalOnWebApplication
@ConditionalOnProperty(name = "shiro.web.enabled", matchIfMissing = true)
public class ShiroWebFilterConfiguration extends AbstractShiroWebFilterConfiguration {

    @Bean
    @Override
    protected ShiroFilterFactoryBean shiroFilterFactoryBean() {
        ShiroFilterFactoryBean filterFactoryBean = super.shiroFilterFactoryBean();

        Map<String, Filter> filters = shiroFilters();
        filterFactoryBean.setFilters(filters);

        return filterFactoryBean;
    }

    @Bean
    protected Map<String, Filter> shiroFilters() {
        Map<String, Filter> filters = new HashMap<>();
        filters.put("filter", myFilter());
    // ...
        return filters;
    }

    @Bean
    protected MyFilter myFilter() {
        return new MyFilter();
    }

    @Bean(name = "filterShiroFilterRegistrationBean")
    protected FilterRegistrationBean filterShiroFilterRegistrationBean() throws Exception {

        FilterRegistrationBean filterRegistrationBean = new FilterRegistrationBean();
        filterRegistrationBean.setFilter((AbstractShiroFilter) shiroFilterFactoryBean().getObject());
        filterRegistrationBean.setOrder(1);

        return filterRegistrationBean;
    }

}
```
I assume there is an registration issue, cause when I autowire some metrics manually ...
```
@Autowired
private MeterRegistry aaa;

@Autowired
private JvmMemoryMetrics bbb;

...

bbb.bindTo(aaa);
```
... they are suddenly seen in actuator. @izeye, @logemann, any idea?

Thanks in advance for any help.

**Comment by izeye** [comment]
@andrejleitner Sorry for the late reply. If you're still suffering from the mentioned issue and could provide a minimal but complete sample on GitHub, I can try to take a look into it although I'm not familiar with Apache Shiro.

**Comment by Wubbwubb** [comment]
I think I've found a possible reason for the problem.

In my application I used an autowired `RestTemplateBuilder` in some of my `Service` classes. The `RestTemplateAutoConfiguration` caused the trouble with the AutoConfiguration order, since the `PrometheusMetricsExportAutoConfiguration` was called by the `RestTemplateAutoConfiguration` (before `MetricsAutoConfiguration`).

After I removed the autowired `RestTemplateBuilder`s the `PrometheusMetricsExportAutoConfiguration` has been configured after `MetricsAutoConfiguration` - like the annotation `@AutoConfigureAfter` in `PrometheusMetricsExportAutoConfiguration` says...
